keps/78-dynamically-reclaiming-resources/README.md

```diff
@@ -0,0 +1,169 @@
+This is not efficient as the Jobs might have different needs during execution, for instance the needs of a `batch/job` having the `parallelism` equal to `completions` will decrease with every Pod finishing it's execution.
+
+### Goals
+
+- Utilize the unused resources of the successfully completed Pods of a running Job.
+
+### Non-Goals
+
+- Preempting a Workload or Pods of a Workload to free resources for a pending Workload.
+- Partially admitting a Workload.
+- Monitor the Job's pod execution.
+
+## Proposal
+
+Reclaiming the resources of the succeeded Pods of a running Job as soon as the Pod completes its execution.
+
+We propose to add a new field `.status.ReclaimablePods` to the Workload API. `.status.ReclaimablePods` is a list that holds the count of Pods belonging to a PodSet whose resources are no longer needed and could be reclaimed by Kueue.
+
+### Pod Completion Accounting
+
+Since the ability of actively monitoring pods execution of a job is not in the scope of the core Kueue implementation, the number of pods for which the resources are no longer needed for each PodSet should be reported by each framework specific `GenericJob` implementation.
+
+For this purpose the `GenericJob` interface should be changed to either:
+
+1. Add an additional method able to report this
+
+```go
+type GenericJob interface {
+    // ...
+
+   // Get reclaimable pods.
+   ReclaimablePods() map[string]int
+}
+
+```
+
+2. Modify the current `Finished` method
+
+```go
+type GenericJob interface {
+   // ...
+
+   // Get reclaimable pods.
+   Finished() (condition metav1.Condition, reclaimablePods map[string]int)
+}
+
+```
+
+#### Example (`batch/job`)
+
+Having a job defined with **P** `parallelism` and **C** `completions`, and **n** number of completed pod executions,
+the expected result should be:
+
+```go
+map[string]int{
+    "main": P - max(P,(C-n))
+}
+```
+### API
+
+A new field `ReclaimablePods` is added to the `.status` of Workload API.
+
+```go
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+
+    // ...
+
+    // list of count of Pods of a PodSet with reclaimable resources 
+    // +optional
+    ReclaimablePods []ReclaimablePod `json:"reclaimablePod"`
+}
+
+// ReclaimedPodSetPod defines the PodSet name and count of Pods
+// belonging to the PodSet whose resources can be reclaimed.
+type ReclaimablePods struct{
+    Name string
+    Count int
+}
+```
+
+## Implementation
+
+### Workload
+#### API
+
+- Add the new field in the workload's status.
+- Validate the data in `status.ReclaimablePods`
+  1. The names must be found in the `PodSets`.
+  2. The cont should never exceed the `PodSets` count.
+  2. The cont should not decrease if the workload is admitted.
+
+
+#### `pkg/workload`
+
+Rework the way `Info.TotalRequests` in computed in order to take the `ReclaimablePods` into account.
+
+#### Controller
+
+Should update the `status.Admission` when `status.ReclaimablePods` changes.
```

Comment:
uhm... that would be 2 API calls in the same object. Is it possible to do 1? Although that means that the kueue-job controller needs to know how to calculate usage, or have access to the workload.Info

Reply:
we might be able to to skip this step, I still need to check how the implementation will go.

Reply:
we can do it without this.